### DISCUSSION TITLE
Create native executable for macos

### DISCUSSION THREAD

**Discussion by mikeptweet** [Moddable-OpenSource/moddable#974] [Q&A] [answered]
Is it possible to create a native executable using this SDK? I want to create a simple websocket client in Javascript that runs in a mac terminal.

**Comment by phoddie** [comment, marked as answer]
Welcome!

(FYI – I converted this question from an Issue report to a Discussion thread)

The Moddable SDK is primarily for building executables for microcontrollers, not computers. We do build and run executables on computers for our simulator (mcsim) and for many of our tools (mcconfig is written in JavaScript and executed by our XS JavaScript engine). We have limited support for building command line tools.

A good place to start is the the JavaScript REPL example at $MODDABLE/examples/js/repl. To run that on macOS, set-up the Moddable SDK and then:

```
cd $MODDABLE/examples/js/repl
mcconfig -m -p x-cli-mac
repl
```
Note a couple things:

- The build target is "x-cli-mac" instead of "mac" or "sim" to indicate that this will run on the command line
- This is a release build (no "-d") to avoid having debugger output in the REPL
- On some versions of macOS you may need to remove `-lpng16` from the the [x-cli-mac makefile](https://github.com/Moddable-OpenSource/moddable/blob/2569d0104769b9769cf2cc071a59fac00dfe8233/tools/mcconfig/make.x-cli-mac.mk#L106) or you will get a linker error (yes, we should fix that)

When it starts up, you should see something like:

```
Moddable JavaScript REPL v0.0.5
  XS engine v13.2.0
  ? for help

> 
```

Use `.help` to see the list of installed commands. You can use the `CLI` class to add commands to the REPL.

Here's the bad news: the REPL does not currently support network operations. Adding that is possible, but not trivial. The REPL does support network operations on embedded devices, like the ESP32 MCU. There is an [outstanding PR](https://github.com/Moddable-OpenSource/moddable/pull/528) from @cmidgley that might be helpful here.

**Reply by mikeptweet** [reply]
Thanks for the detailed reply ... too bad about the lack of networking support.

**Reply by phoddie** [reply]
If I may ask, what is it that you are looking to accomplish -- why use the Moddable SDK for a WebSocket-based command line tool?

**Comment by mikeptweet** [comment]
I am looking for something lightweight in terms of footprint (memory, cpu, etc) that I can distribute as a single executable with no dependencies.

**Reply by phoddie** [reply]
That's a neat idea. Code written for microcontrollers does need to be both small and dependency free. We do distribute binaries of our simulator to clients regularly, and it works really well. 

Did you take a look at the PR from @cmidgley? It seems like it might be what you need. If it is (or is close enough), we could re-open discussion on what it would mean to integrate that

If it would be easier to discuss in real-time, I'm often available on our [Gitter](https://gitter.im/embedded-javascript/moddable).

**Comment by cmidgley** [comment]
Hey all - my original need was very similar, but for a different reason.  I needed to have a program that would run on a microcontroller as well as a PC/Mac/Linux and it made for a single source code solution for all of them.  I ended up going another direction - making a build system where a build can target Node or Moddable and they can share NPMs (it's a complex and custom build setup, dependent on things like Windows using WSL, Rush monorepo, custom Jest testing ... very opinionated and specific to my needs, so not easily shared otherwise I would!)  

Of note, the PR also supports (supported?) mods.  I may have a new need for that PR again, in which case I might crack it open again (it claims it can merge without conflict, so at least that's a start).

**Reply by phoddie** [reply]
Thanks, @cmidgley! I thought the request sounded familiar. Good ideas just keep coming back, I guess. ;)
